components/heap/heap_caps.c

```diff
@@ -77,6 +85,54 @@ esp_err_t heap_caps_register_failed_alloc_callback(esp_alloc_failed_hook_t callb
     return ESP_OK;
 }
 
+IRAM_ATTR static bool heap_caps_low_memory_check( size_t size, uint32_t caps)
+{
+#if CONFIG_HEAP_LOW_MEMORY_CALLBACK_COUNT
+    for (int i = 0; i < CONFIG_HEAP_LOW_MEMORY_CALLBACK_COUNT; i++) {
+        heap_low_memory_callback_t ith = low_memory_callbacks[i];
+        if (ith.callback && (ith.caps & caps) > 0) {
+            size_t free = heap_caps_get_free_size(ith.caps);
+            if (free < ith.threshold + size) {
+                return ith.callback(caps, size, free);
+            }
+        }
+    }
+#endif
+    return true;
+}
+
+esp_err_t heap_caps_register_low_heap_callback(uint32_t caps, uint32_t threshold, esp_heap_low_memory_hook_t callback)
+{
+    if (callback == NULL) {
+        return ESP_ERR_INVALID_ARG;
+    }
+
+    // exists?
+    for (int i = 0; i < CONFIG_HEAP_LOW_MEMORY_CALLBACK_COUNT; i++) {
+        if (low_memory_callbacks[i].caps == caps) {
+            low_memory_callbacks[i].caps = caps;
+            low_memory_callbacks[i].threshold = threshold;
+            low_memory_callbacks[i].callback = callback;
+            return ESP_OK;
+        }
+    }
+
+    // add
+    for (int i = 0; i < CONFIG_HEAP_LOW_MEMORY_CALLBACK_COUNT; i++) {
+        if (low_memory_callbacks[i].callback == NULL) {
+            low_memory_callbacks[i].caps = caps;
+            low_memory_callbacks[i].threshold = threshold;
+            low_memory_callbacks[i].callback = callback;
+            return ESP_OK;
+        }
+    }
```

Comment:
Wouldn't it be more efficient to make only 1 loop looking for existing entry and adding one entry if no entry found? Also, you are using the register function to modify an existing entry when `low_memory_callbacks[i].caps == caps` (here we want to update the `threshold` and/or `callback`) but it would be nice to also allow the user to call this function to update the `threshold` and/or `caps` by checking for `low_memory_callbacks[i].callback == callback`.

Reply:
As for 1 loop vs 2, I think we should do whatever is most readable. There are no perf concerns either way imo.

For the first loop, yous suggest this? Yes that sounds good.
`if (low_memory_callbacks[i].caps == caps || low_memory_callbacks[i].callback == callback) {`